### PR TITLE
Throw exception when using convention mapping with all new `Property` types

### PR DESCRIPTION
**Pull request by lacasseio**
### Context

See https://github.com/gradle/gradle/issues/8404

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=lacasseio%2Fdeprecate-convention-mapping-for-provider)
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java [hunk 1]
```diff
@@ -38,15 +46,28 @@
...
     private final IConventionAware _source;
     private final Set<String> _propertyNames;
+    private final Map<String, Boolean> _propertyNamesAsProviderMapping;
     private final Map<String, MappedPropertyImpl> _mappings = new HashMap<String, MappedPropertyImpl>();
 
     public ConventionAwareHelper(IConventionAware source, Convention convention) {
         this._source = source;
         this._convention = convention;
+        this._propertyNamesAsProviderMapping = ClassInspector.inspect(source.getClass()).getProperties().stream().collect(Collectors.toMap(it -> it.getName(), it -> isProviderType(it.getGetters())));
         this._propertyNames = JavaPropertyReflectionUtil.propertyNames(source);
```
Comment (lacasseio) [inline subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java]: ❓ Not sure if those two should be merged together.

File: subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java [hunk 2]
```diff
@@ -38,15 +46,28 @@
...
     }
 
+    private static boolean isProviderType(List<Method> getters) {
+        return getters.stream().anyMatch(ConventionAwareHelper::isProviderType);
+    }
+
+    private static boolean isProviderType(Method getter) {
+        return Property.class.isAssignableFrom(getter.getReturnType()) || ListProperty.class.isAssignableFrom(getter.getReturnType()) || SetProperty.class.isAssignableFrom(getter.getReturnType()) || MapProperty.class.isAssignableFrom(getter.getReturnType());
```
Comment (adammurdoch) [inline subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java]: I think this should be just `Provider.class.isAssignableFrom(getter.returnType)`. We don't want convention mappings to be applied to any kind of `Provider`.

File: subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java (hunk 1 above)
Comment (adammurdoch) [reply subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java]: Definitely. I think this is going to kill performance, as we're reflecting on the whole type hierarchy for every instance created, without caching the result. Plus we're doing the same inspection again in class decoration (and again just after the task is created, if this happens to represent a task ). We shouldn't add another place where we visit the type hierarchy, so at the very least we should combine the property name and property type calculations here, then later we can look at how we merge the other places.

File: subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java (hunk 2 above)
Comment (lacasseio) [reply subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java]: The convention mapping works for `Provider` but is completely ignored on `Property` types. This PR is strictly addressing the `Property` and throws an exception. For the other type, I was going to issue a deprecation warning as it works properly but we want to stop people from using them as a separate PR.

**Comment by lacasseio** [comment]
@adammurdoch This is just a first step. I was going to address `Provider` and the others as a separate PR.

I agree with:
> We should also consider disallowing mutable properties with these types, too. That is, disallow having setters for these properties. It doesn't make sense to allow changing the property value using a setter but disallowing it using a convention mapping.

Let's open another issue and address this as part of another PR.

**Comment by lacasseio** [comment]
I saw the comment on the issue, I add the `Provider` case in this PR.

**Comment by adammurdoch** [comment]
I looked at the PR and the issue in the wrong order, I guess. It's fine to just have a deprecation warning for `Provider` types if it currently works and people are using it, and just fail for `Property` types for now. Or perhaps let's just treat everything the same and only deprecate this pattern for all `Provider` and property types rather than deprecating some and failing for others.

**Comment by lacasseio** [comment]
Closing to reflect that we will address this differently for 5.4

**Pull request closed without merging**